3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/RefererParserEnrichment.scala

```diff
@@ -101,9 +119,22 @@ case class RefererParserEnrichment(
    * @return a Tuple3 containing referer medium,
    *         source and term, all Strings
    */
-  def extractRefererDetails(uri: URI, pageHost: String): Option[Referer] =
-    for {
-      r <- RefererParser.parse(uri, pageHost, domains)
-      t = r.term.flatMap(t => CU.fixTabsNewlines(t))
-    } yield termLens.set(r, t)
+  def extractRefererDetails(uri: URI, pageHost: String): EitherT[IO, Exception, Option[Referer]] = {
+    val fixedURI = if (includeSchemas) new URIBuilder(uri.toString).setScheme("http").build else uri
+    val io: EitherT[IO, Exception, Option[Referer]] = for {
+      parser <- EitherT(Parser.create[IO](getClass.getResource(referersJsonPath).getPath))
+      r <- EitherT
+        .fromOption[IO](parser.parse(fixedURI, Some(pageHost), domains), new Exception("No parseable referer"))
```

Comment:
Actually the Exception is mostly for reference, as it's not thrown, and is actually dropped in the same step by mapping over the `EitherT`. If we wanted to we could, for example, log it. 

The main reason it's included here is for agreement with the `EitherT` structure that `referer-parser` expects in case there's an I/O problem. `Parser.create` returns `IO[Either[Exception, Parser]]` and `EitherT` wraps that to reduce boilerplate when mapping, which allows us to, as we do here, pull the Parser directly out in a for-comprehension without having to deal with the Either.

The point being, the Exception here is incidental, and throwaway, and is just a product of the way the IO is invoked. We don't actually care about it because if there's an Exception that just means we don't have a referer, and we're counting this URI as `direct`. If we were concerned about *which* URIs we're throwing out as invalid, we could add a logging step here, but that seems like overkill here, and would require some investigation into how and where Snowplow logs, since I'm not familiar with that.

Reply:
I should probably also point out, to answer your question, that in line 114, the `getResource()` call supplies its own Exception if nothing usable is found at the path. In line 116 we're supplying a new Exception that will indicate an invalid referer URI, mostly just because one is required for the function invocation. I'll update the Exception message to more clearly indicate that.